tools/nl/nodejs_query_differ/differ.py

```diff
@@ -30,82 +37,120 @@
+  if not test_folder:
     blobs = bucket.list_blobs(prefix=FLAGS.env)
     # Get all the dates where there have been cron testing runs done
     dates = set()
     for b in blobs:
       date = b.name[len(FLAGS.env) + 1:].split('/')[0]
-      dates.add(date)
+      if date != _GOLDEN_FOLDER:
+        dates.add(date)
     sorted_dates = sorted(dates, reverse=True)
-    if len(sorted_dates) < 2:
+    if len(sorted_dates) < 1:
       return
-    # set base_folder and test_folder if they weren't passed in as flags
-    if not base_folder:
-      base_folder = sorted_dates[1]
-    if not test_folder:
-      test_folder = sorted_dates[0]
-  return (base_folder, test_folder)
-
-
-def run_diff(base_blobs, test_blobs, output_file):
-  """Gets the diff between the base blobs and test blobs and output it"""
-  base_jsons = {}
-  for b in base_blobs:
+    test_folder = sorted_dates[0]
+  return test_folder
+
+
+def get_diff(golden_blobs, test_blobs):
+  """Gets the results of diffing between the base blobs and test blobs"""
+  golden_jsons = {}
+  for b in golden_blobs:
     key = b.name.split('/')[-1].split('.')[0]
-    base_jsons[key] = json.loads(b.download_as_string())
+    golden_jsons[key] = json.loads(b.download_as_string())
   test_jsons = {}
   for t in test_blobs:
     key = t.name.split('/')[-1].split('.')[0]
     test_jsons[key] = json.loads(t.download_as_string())
-  report = {}
-  for key, base_json in base_jsons.items():
+
+  results = {}
+  for key, golden_json in golden_jsons.items():
     if not key in test_jsons:
-      report[key] = "Missing in test folder"
+      results[key] = "Missing in test folder"
+      continue
     test_json = test_jsons[key]
-    del base_json['debug']
     del test_json['debug']
-    diff = DeepDiff(base_json, test_json)
+    diff = DeepDiff(golden_json, test_json)
     if diff:
-      report[key] = json.dumps(diff.to_json())
+      results[key] = json.dumps(diff.to_json())
     else:
-      report[key] = 'Success'
-  # TODO: have a UI for showing the result
-  with open(output_file, 'w') as f:
-    f.write(json.dumps(report))
-  print("Done running differ")
-  print(f'Diff report saved locally to: {_OUTPUT_FILE}')
+      results[key] = _DIFF_SUCCESS_MSG
+  return results
+
+
+def output_results(results, gcs_bucket) -> str:
+  """Outputs results either to gcs or locally"""
+  if FLAGS.gcs_output_folder:
+    gcs_filename = f'{FLAGS.gcs_output_folder}/{_OUTPUT_FILE}'
+    blob = gcs_bucket.blob(gcs_filename)
+    with blob.open('w') as f:
+      f.write(json.dumps(results))
+    results_path = f'gs://{_GCS_BUCKET}/{gcs_filename}'
+    logging.info(f'Diff results saved to gcs path: {results_path}')
+    return results_path
+  else:
+    with open(_OUTPUT_FILE, 'w') as f:
+      f.write(json.dumps(results))
+    logging.info(f'Diff results saved locally to: {_OUTPUT_FILE}')
+    return _OUTPUT_FILE
+
+
+def output_failure_email(results, golden_folder, test_folder, results_path):
```

Comment:
Do you have an example failure diff result?

Reply:
I have a screenshot in the PR description?

Reply:
I was looking for the file contents here.

Found it from your screenshot image, after scrolling for a bit:

`{"bar": "Success", "disaster": "Success", "map_rank": "Success", "scatter": "Success", "scatter_non_bard": "Success", "timeline": "Success", "timeline_all": "Success"}`

Can we do indent=2?

Also, how does a failure look?

Reply:
Hmm, ok - what would the next step to debug be - regenerate goldens and review?

Reply:
Might be helpful to have a paragraph on "steps to take on diff failure" in go/dcrelease or somewhere easily accessible.